File: app/src/animation/ AnimatedSprite.java
package src.animation;

import java.awt.*;
import java.awt.geom.AffineTransform;
import java.awt.image.BufferedImage;
import java.util.HashMap;

public class AnimatedSprite
{
    public final static String IDLE = "idle";
    public final static String WALK = "walk";

    private final HashMap<String, BufferedImage[]> sprites;
    private final float frameLength;
    private String state;
    private float time;
    private int animFrame;
    private boolean rotation;

    public AnimatedSprite(String startState, BufferedImage[] startAnimation, float frameLength)
    {
        this.sprites = new HashMap<String, BufferedImage[]>();
        this.sprites.put(startState, startAnimation);
        this.state = startState;
        this.frameLength = frameLength;
        this.time = 0;
        this.animFrame = 0;
    }

    public void addState(String state, BufferedImage[] animation)
    {
        sprites.put(state, animation);
    }

    public void setRotation(boolean lookToRight)
    {
        rotation = lookToRight;
    }

    public boolean getRotation(){ return rotation; }

    public void setState(String state)
    {
        if(!sprites.containsKey(state))
            throw new IllegalArgumentException("This animated sprite doesn't have this state");

        this.state = state;
    }

    public void update(float dt)
    {
        time+=dt;
        if(time > frameLength)
        {
            animFrame++;
            time = time - frameLength;
        }
    }

    public BufferedImage getSprite()
    {
        BufferedImage[] images = sprites.get(state);
        return applyRotation(images[animFrame % images.length], rotation);
    }s
}

File: app/src/animation/ AnimatedSprite.kt
package src.animation

import java.awt.geom.AffineTransform
import java.awt.image.BufferedImage
import java.util.HashMap

class AnimatedSprite(
    startState: String,
    startAnimation: Array<BufferedImage>,
    frameLength: Float
) {
    private val sprites: HashMap<String, Array<BufferedImage>>
    private val frameLength: Float
    private var state: String
    private var time: Float
    private var animFrame: Int

    var rotation = false
    fun addState(state: String?, animation: Array<BufferedImage>) {
        sprites.put(state, animation)
    }

    fun setState(state: String) {
        if (!sprites.containsKey(state)) throw IllegalArgumentException("This animated sprite doesn't have this state")
        this.state = state
    }

    fun update(dt: Float) {
        time += dt
        if (time > frameLength) {
            animFrame++
            time = time - frameLength
        }
    }

    val sprite: BufferedImage
        get() {
            val images: Array<BufferedImage> = sprites.get(state)
            return applyRotation(images[animFrame % images.size], rotation)
        }

    companion object {
        const val IDLE = "idle"
        const val WALK = "walk"
    }

    init {
        sprites = HashMap<String, Array<BufferedImage>>()
        sprites.put(startState, startAnimation)
        state = startState
        this.frameLength = frameLength
        time = 0f
        animFrame = 0
    }
}